packages/backend/src/workers/unstaker.ts

```
@@ -0,0 +1,394 @@
+        console.log(
+          `UNSTAKE tx was NOT sent for app ${app.name} ${address}! This is an issue with the provider node and pocketJS.`
+        )
+        return
+      }
+
+      app.status = APPLICATION_STATUSES.AWAITING_UNSTAKING
+      app.updatedAt = new Date(Date.now())
+
+      ctx.logger.log(
+        `fillAppPool(): sent funds to account ${address} on tx ${txHash}`
+      )
+      await app.save()
+    })
+  )
+}
+
+export async function unstakeApps(ctx): Promise<void> {
+  const apps = await Application.find({
+    status: APPLICATION_STATUSES.AWAITING_UNSTAKING,
+  }).limit(10)
+
+  Promise.allSettled(
+    apps.map(async function unstake(app) {
+      const updatedAt = dayjs(app.updatedAt)
+      const now = dayjs()
+
+      // For TX propagation purposes, we wait a bit in between sending an unstake tx and a stake tx.
+      // This way they're included in the same (or subsequent) block, and we don't leave the gap open for too long.
+      const isGracePeriodOver = now.diff(updatedAt, 'minute') >= 10
+
+      if (!isGracePeriodOver) {
+        return
+      }
+
+      // @ts-ignore
+      await unstakeApplication(app, ctx)
+    })
+  )
+}
+
+export async function stakeAppsForSlots(ctx): Promise<void> {
+  const apps = await Application.find({
+    status: APPLICATION_STATUSES.AWAITING_SLOT_STAKING,
+  }).limit(10)
+
+  Promise.allSettled(
+    apps.map(async function unstake(app) {
+      const updatedAt = dayjs(app.updatedAt)
+      const now = dayjs()
+
+      // For TX propagation purposes, we wait a bit in between sending an unstake tx and a stake tx.
+      // This way they're included in the same (or subsequent) block, and we don't leave the gap open for too long.
+      const isGracePeriodOver = now.diff(updatedAt, 'minute') >= 5
+
+      if (!isGracePeriodOver) {
+        return
+      }
+
+      // @ts-ignore
+      await stakeApplication(app, ctx)
+    })
+  )
+}
+
+export async function removeFundsFromApps(ctx): Promise<void> {
```

My concern with this complexity and multiple steps is with the visibility.

Can we build in better visibility rather than checking logs?

---

What would classify as better visibility? Open to ideas—agree that manually checking logs can be a bit cumbersome.

---

I'd like to see some alerts that bubble up to us proactively rather than us having to read logs retroactively.

Maybe something with DataDog?  If your logs end up there (they probably already do), you could work with @JonathonJulian to create alerts based on logged error conditions.

---

Sounds good! Will ping Jon to see what we can do.